File: internal/continuous_integration/short/plotter.py
#!/tool/pandora64/.package/python-3.8.0/bin/python3

import dictionary
import log
#import matplotlib.pyplot
import numpy
import csv
import os
import subprocess
import sys

class Plot():
    def __init__(self, args, archives, changeset_to_build):
        self._args = args
        self._archives = archives
        self._changelog = dictionary.ChangelogToMostRecentBuild()
        self._changelog.generate(changeset_to_build)
        print(self._changelog.dump())

    def abbreviate_changesets(self, changesets):
        return [changeset[0:8] for changeset in changesets]

    @staticmethod
    def write_dict_to_file(tracker, field_names, file_name):
        with open(file_name, 'w') as csvfile:
            writer = csv.DictWriter(csvfile, fieldnames=field_names)
            writer.writeheader()
            writer.writerows(tracker)
    
    @staticmethod
    def check_and_add_to_dict(dictionary, key, array):
        if len(array) > 0:
            dictionary[key] = array[0]
        else:
            dictionary[key] = 0

    def changeset_slice(self):
        self._log_tracker = log.Tracker(self._args, self._archives)
        for changeset in self._changelog._all_changesets:
            if changeset in self._changelog.data.keys():
                build_id = self._changelog.data[changeset]
                self._log_tracker.add(changeset, build_id)
        print(self._log_tracker.dump())
        
        """
        separate out dictionaries based on operation
        and prepare them in a format that works with
        the csv module
        """
        put_tracker = []
        put_nbi_tracker = []
        get_tracker = []
        get_nbi_tracker = []
        amo_tracker = []
        ping_pong_tracker = []
        prev_commit = list(self._log_tracker._data.keys())[0][0]
        amo_dict = {}
        for key, value in self._log_tracker._data.items():
            if (key[1] == "put.log"):
                put_tracker.append({'Commit':key[0][0:7],
                                    'b1':value.latency[0],
                                    'b2':value.latency[1],
                                    'b4':value.latency[2],
                                    'b8':value.latency[3],
                                    'b16':value.latency[4],
                                    'b32':value.latency[5],
                                    'b64':value.latency[6],
                                    'b128':value.latency[7],
                                    'b256':value.latency[8],
                                    'b512':value.latency[9],
                                    'b1024':value.latency[10],
                                    'b2048':value.latency[11],
                                    'b4096':value.latency[12],
                                    'b8192':value.latency[13],
                                    'b16384':value.latency[14],
                                    'b32768':value.latency[15]
                                    })
            if (key[1] == "put_nbi.log"):
                put_nbi_tracker.append({'Commit':key[0][0:7],
                                    'b1':value.latency[0],
                                    'b2':value.latency[1],
                                    'b4':value.latency[2],
                                    'b8':value.latency[3],
                                    'b16':value.latency[4],
                                    'b32':value.latency[5],
                                    'b64':value.latency[6],
                                    'b128':value.latency[7],
                                    'b256':value.latency[8],
                                    'b512':value.latency[9],
                                    'b1024':value.latency[10],
                                    'b2048':value.latency[11],
                                    'b4096':value.latency[12],
                                    'b8192':value.latency[13],
                                    'b16384':value.latency[14],
                                    'b32768':value.latency[15]
                                    })
            if (key[1] == "get.log"):
                get_tracker.append({'Commit':key[0][0:7],
                                    'b1':value.latency[0],
                                    'b2':value.latency[1],
                                    'b4':value.latency[2],
                                    'b8':value.latency[3],
                                    'b16':value.latency[4],
                                    'b32':value.latency[5],
                                    'b64':value.latency[6],
                                    'b128':value.latency[7],
                                    'b256':value.latency[8],
                                    'b512':value.latency[9],
                                    'b1024':value.latency[10],
                                    'b2048':value.latency[11],
                                    'b4096':value.latency[12],
                                    'b8192':value.latency[13],
                                    'b16384':value.latency[14],
                                    'b32768':value.latency[15]
                                    })
            if (key[1] == "get_nbi.log"):
                get_nbi_tracker.append({'Commit':key[0][0:7],
                                    'b1':value.latency[0],
                                    'b2':value.latency[1],
                                    'b4':value.latency[2],
                                    'b8':value.latency[3],
                                    'b16':value.latency[4],
                                    'b32':value.latency[5],
                                    'b64':value.latency[6],
                                    'b128':value.latency[7],
                                    'b256':value.latency[8],
                                    'b512':value.latency[9],
                                    'b1024':value.latency[10],
                                    'b2048':value.latency[11],
                                    'b4096':value.latency[12],
                                    'b8192':value.latency[13],
                                    'b16384':value.latency[14],
                                    'b32768':value.latency[15]
                                    })
            if (key[1] == "ping_pong.log"):
                ping_pong_tracker.append({'Commit':key[0][0:7],
                                          'latency':value.latency[0]
                                         })

            # check to see if we have moved to a new commit
            # if we have, store the dict in the amo_tracker
            if (key[0] != prev_commit):
                amo_dict['Commit'] = prev_commit[0:7]
                amo_tracker.append(amo_dict.copy())
                amo_dict.clear()
            
            prev_commit = key[0]
            
            if (key[1] == "amo_add.log"):
                self.check_and_add_to_dict(amo_dict, 'add', value.latency)
            if (key[1] == "amo_cswap.log"):
                self.check_and_add_to_dict(amo_dict, 'cswap', value.latency)
            if (key[1] == "amo_fadd.log"):
                self.check_and_add_to_dict(amo_dict, 'fadd', value.latency)
            if (key[1] == "amo_fcswap.log"):
                self.check_and_add_to_dict(amo_dict, 'fcswap', value.latency)
            if (key[1] == "amo_fetch.log"):
                self.check_and_add_to_dict(amo_dict, 'fetch', value.latency)
            if (key[1] == "amo_finc.log"):
                self.check_and_add_to_dict(amo_dict, 'finc', value.latency)
            if (key[1] == "amo_inc.log"):
                self.check_and_add_to_dict(amo_dict, 'inc', value.latency)
        
        # store the last commit's amo data
        amo_dict['Commit'] = prev_commit[0:7]
        amo_tracker.append(amo_dict.copy())

        # write put results into a file:
        size_field_names= ['Commit','b1','b2','b4','b8','b16','b32','b64','b128','b256','b512','b1024','b2048','b4096','b8192','b16384','b32768']
        amo_field_names= ['Commit','add','cswap','fadd','fcswap','fetch','finc','inc']
        ping_pong_field_names= ['Commit','latency']

        self.write_dict_to_file(put_tracker, size_field_names, "put.csv")
        self.write_dict_to_file(put_nbi_tracker, size_field_names, "put_nbi.csv")
        self.write_dict_to_file(get_tracker, size_field_names, "get.csv")
        self.write_dict_to_file(get_nbi_tracker, size_field_names, "get_nbi.csv")
        self.write_dict_to_file(amo_tracker, amo_field_names, "amo.csv")
        self.write_dict_to_file(ping_pong_tracker, ping_pong_field_names, "ping_pong.csv")

        # make a directory and execute the R script to generate plots in that directory
        current_dir = os.getcwd()
        plot_dir = os.path.join(current_dir, 'plots')
        if not os.path.exists(plot_dir):
            os.makedirs(plot_dir)

        changeset_a = list(self._log_tracker._data.keys())[0][0]
        changeset_b = list(self._log_tracker._data.keys())[-1][0]

        # check if the provided changesets are correct
        if (self._args.changeset_range):
            found_changeset_a = False
            found_changeset_b = False
            for key, value in self._log_tracker._data.items():
                if (found_changeset_a and found_changeset_b):
                    break
                if (not found_changeset_a):
                    if (self._args.changeset_range[0] == key[0]):
                        found_changeset_a = True
                if (not found_changeset_b):
                    if (self._args.changeset_range[1] == key[0]):
                        found_changeset_b = True
            
            if ((not found_changeset_a) and (not found_changeset_b)):
                sys.exit("One of the specified changesets was not found. Please specify correct/complete commit IDs.")
            else:
                changeset_a = self._args.changeset_range[0]
                changeset_b = self._args.changeset_range[1]

        r_command = "Rscript ./plotter.R -o ./plots -a " + changeset_a[0:7] + " -b " + changeset_b[0:7]

        print(r_command)
        subprocess.check_call(r_command, shell=True)


    def one_changeset_plot(self):
        found_changeset = 0
        non_amo_tracker = []
        amo_tracker = []
        ping_pong_tracker = []
        for key, value in self._log_tracker._data.items():
            if (key[0] == self._args.one_changeset):
                found_changeset = 1
                if (key[1] == "put.log"):
                    put_vals = value.latency
                if (key[1] == "put_nbi.log"):
                    put_nbi_vals = value.latency
                if (key[1] == "get.log"):
                    get_vals = value.latency
                if (key[1] == "get_nbi.log"):
                    get_nbi_vals = value.latency
                if (key[1] == "amo_add.log"):
                    amo_tracker.append({'op':'add',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_add.log"):
                    amo_tracker.append({'op':'add',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_cswap.log"):
                    amo_tracker.append({'op':'cswap',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_fadd.log"):
                    amo_tracker.append({'op':'fadd',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_fcswap.log"):
                    amo_tracker.append({'op':'fcswap',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_fetch.log"):
                    amo_tracker.append({'op':'fetch',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_finc.log"):
                    amo_tracker.append({'op':'finc',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "amo_inc.log"):
                    amo_tracker.append({'op':'inc',
                                        'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })
                if (key[1] == "ping_pong.log"):
                    ping_pong_tracker.append({'latency': value.latency[0] if len(value.latency) > 0 else 0
                                        })


        if (not found_changeset):
            sys.exit("The requested changeset was not found. Please specify correct/complete commit IDs.")
        
        index = 0
        for size in [1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768]:
            non_amo_tracker.append({'size':size,
                                    'put':put_vals[index],
                                    'put_nbi':put_nbi_vals[index],
                                    'get':get_vals[index],
                                    'get_nbi':get_nbi_vals[index]
                                    })
            index = index + 1
        
        # write results into a file:
        non_amo_field_names= ['size','put','put_nbi','get','get_nbi']
        amo_field_names= ['op','latency']
        ping_pong_field_names= ['latency']

        self.write_dict_to_file(non_amo_tracker, non_amo_field_names, "non_amo_one_changeset.csv")
        self.write_dict_to_file(amo_tracker, amo_field_names, "amo_one_changeset.csv")
        self.write_dict_to_file(ping_pong_tracker, ping_pong_field_names, "ping_pong_one_changeset.csv")

        # call the R script with an option that tells it to plot figures for
        r_command = "Rscript ./plotter.R -o ./plots -c " + self._args.one_changeset

        print(r_command)
        subprocess.check_call(r_command, shell=True)

File: internal/continuous_integration/short/plotter.R
#!/usr/bin/env Rscript

# load the required libraries:
library(tidyverse)
library(RColorBrewer)
library(optparse)

# declare some helper functions
ggpreview <- function (..., device = "png") {
  fname <- tempfile(fileext = paste0(".", device))
  ggplot2::ggsave(filename = fname, device = device, ...)
  system2("open", fname)
  invisible(NULL)
}

set_right_order <- function(df) {
  # reverse the order of the rows so that oldest commit is first
  df <- df %>% map_df(rev)
  # ensure that ggplot plots the x-axis in the right order
  df$Commit <- factor(df$Commit, levels = unique(df$Commit))
  return(df)
}

plot_and_save <- function(df, xval, yval, title, subtitle, xlabel, filename) {
  p <- ggplot(df, aes_string(x=xval, y=yval, group=1)) +
    geom_line(size = 0.5, color=mycolors[1]) +
    geom_point(size = 1.5, alpha = 1, color=mycolors[2]) +
    theme_minimal() +
    expand_limits(y=0) +
    xlab(xlabel) +
    ggtitle(title, subtitle = subtitle) +
    theme(
      axis.text.x = element_text(angle=90,hjust=1),
      axis.title.y = element_blank()
    ) +
  scale_fill_manual(values = mycolors)
  #ggpreview(width=7.5, height=5, units="in", dpi=500)
  ggsave(filename, p, device=pdf, dpi=500)
}

## Set up options ##

option_list = list(
    make_option(c("-o", "--output"), type="character", default=NULL, action="store",
                   help="path (without trailing /) to a folder that will
                   contain the plots", metavar="folder-path"),
    make_option(c("-a", "--changeset_a"), type="character", default=NULL, action="store",
                   help="beginning (inclusive) changeset of slice", metavar="changeset"),
    make_option(c("-b", "--changeset_b"), type="character", default=NULL, action="store",
                   help="ending (inclusive) changeset of slice", metavar="changeset"),
    make_option(c("-c", "--one_changeset"), type="character", default=NULL, action="store",
                help="if set, will prepare plots for one changeset; if not, plots for a changeset slice")

)

## SCRIPT START ##

# parse the options
opt_parser <- OptionParser(option_list=option_list)
opts <- parse_args(opt_parser)
if (is.null(opts$output)) {
    print_help(opt_parser)
    stop("Please set the --output flag.", call.=FALSE)
}
slice_opt = 0
single_opt = 0
if (!is.null(opts$changeset_a) && !is.null(opts$changeset_b)) {
    slice_opt = 1
}
if (!is.null(opts$one_changeset)) {
    single_opt = 1
}

if ( (slice_opt && single_opt) || (!slice_opt && !single_opt) ) {
    stop("Please supply a slice or a single changeset, not both.", call.= FALSE)
}

# choose color palette
mycolors <- brewer.pal(5, "Set2")

if (length(opts$one_changeset) > 0) {
  ## Plotting data for a single changeset ##

  # read the files
  non_amo   <- read.csv("non_amo_one_changeset.csv", header=TRUE)
  amo       <- read.csv("amo_one_changeset.csv", header=TRUE)
  ping_pong <- read.csv("ping_pong_one_changeset.csv", header=TRUE)

  # ensure that ggplot plots the x-axis in the right order
  non_amo$size <- factor(non_amo$size, levels = unique(non_amo$size))
  amo$op <- factor(amo$op, levels = unique(amo$op))

  # plot
  non_amo_ops <- list("put","put_nbi","get","get_nbi")
  for (op in non_amo_ops) {
    plot_and_save(df=non_amo,
                  xval="size",
                  yval=op,
                  title=op,
                  subtitle="Latency (us)",
                  xlabel="Message size (bytes)",
                  filename=paste(opts$output,"/",op,"_changeset_",opts$one_changeset,".pdf", sep="")
                  )
  }

  # prepare data for plots with fixed message size and ops as x axis
  non_amo$bsize <- paste("b",non_amo$size,sep="") # (so that the columns in non_amo_t start with a character)
  non_amo_t <- setNames(data.frame(t(non_amo[,2:5])), non_amo[,6]) # transpose + set column names
  non_amo_t$op <- colnames(non_amo[,2:5]) # make a column with operation names

  sizes <- colnames(non_amo_t[,-(length(colnames(non_amo_t)))])
  for (size in sizes) {
    plot_and_save(df=non_amo_t,
                  xval="op",
                  yval=size,
                  title=paste(sub('.', '', size),"byte"),
                  subtitle="Latency (us)",
                  xlabel="Operation",
                  filename=paste(opts$output,"/",size,"_changeset_",opts$one_changeset,".pdf", sep="")
    )
  }

  plot_and_save(df=amo,
                xval="op",
                yval="latency",
                title="Atomics",
                subtitle="Latency (us)",
                xlabel="Operation",
                filename=paste(opts$output,"/atomic_changeset_",opts$one_changeset,".pdf", sep="")
  )

  ping_pong$type <- c("ping_pong")
  p<-ggplot(ping_pong, aes(x=type, y=latency, fill=type)) +
    geom_bar(stat="identity", width=0.5) +
    theme_minimal() +
    ggtitle("Ping pong", subtitle = "Latency (us)") +
    theme(
      axis.title.y = element_blank(),
      axis.text.y = element_blank(),
      axis.title.x = element_blank(),
      legend.position = "none"
      ) +
    coord_flip() +
    scale_fill_manual(values = mycolors)
  #ggpreview(width=7.5, height=5, units="in", dpi=500)
  ggsave(paste(opts$output,"/ping_pong_changeset_",opts$one_changeset,".pdf", sep=""), p, device=pdf, dpi=500)

} else {
  ## Plotting across a changeset slice ##

  # read the files
  put             <- read.csv("put.csv", header=TRUE)
  put_nbi         <- read.csv("put_nbi.csv", header=TRUE)
  get             <- read.csv("get.csv", header=TRUE)
  get_nbi         <- read.csv("get_nbi.csv", header=TRUE)
  amo             <- read.csv("amo.csv", header=TRUE)
  ping_pong       <- read.csv("ping_pong.csv", header=TRUE)

  # slice out the commits
  start   <- match(c(opts$changeset_a), put$Commit)
  end     <- match(c(opts$changeset_b), put$Commit)
  # (start and end should be the same for all the frames) #
  put             <- put[start:end,]
  put_nbi         <- put_nbi[start:end,]
  get             <- get[start:end,]
  get_nbi         <- get_nbi[start:end,]
  amo             <- amo[start:end,]
  ping_pong       <- ping_pong[start:end,]

  put             <- set_right_order(put)
  put_nbi         <- set_right_order(put_nbi)
  get             <- set_right_order(get)
  get_nbi         <- set_right_order(get_nbi)
  amo             <- set_right_order(amo)
  ping_pong       <- set_right_order(ping_pong)

  # plot
  non_amo_ops <- list("put","put_nbi","get","get_nbi")
  sizes_to_subtitle_map <- list("b1"="1 byte",
                                "b2"="2 bytes",
                                "b4"="4 bytes",
                                "b8"="8 bytes",
                                "b16"="16 bytes",
                                "b32"="32 bytes",
                                "b64"="64 bytes",
                                "b128"="128 bytes",
                                "b256"="256 bytes",
                                "b512"="512 bytes",
                                "b1024"="1024 bytes",
                                "b2048"="2048 bytes",
                                "b4096"="4096 bytes",
                                "b8192"="8192 bytes",
                                "b16384"="16384 bytes",
                                "b32768"="32768 bytes")
  for (op in non_amo_ops) {
    for (size in names(sizes_to_subtitle_map)) {
      plot_and_save(df=eval(parse(text=op)),
                    xval="Commit",
                    yval=size,
                    title=op,
                    subtitle=paste("Latency (us) for ",sizes_to_subtitle_map[[size]],sep=""),
                    xlabel="Commit (older to newer)",
                    filename=paste(opts$output,"/",op,"_",size,".pdf", sep="")
      )
      }
  }

  amo_ops <- list("add","cswap","fadd","fcswap","fetch","finc","inc")
  for (op in amo_ops) {
    plot_and_save(df=amo,
                  xval="Commit",
                  yval=op,
                  title=op,
                  subtitle="Latency (us)",
                  xlabel="Commit (older to newer)",
                  filename=paste(opts$output,"/",op,".pdf", sep="")
    )
  }

  plot_and_save(df=ping_pong,
                xval="Commit",
                yval="latency",
                title="ping_pong",
                subtitle="Latency (us)",
                xlabel="Commit (older to newer)",
                filename=paste(opts$output,"/","ping_pong.pdf", sep="")
  )
}

## SCRIPT END ##
